ls1intum/Artemis
# `Programming exercises`: Fix an edge case issue in student exercise actions

### Checklist
#### General


#### Client

### Motivation and Context


Sometimes instructors have to postpone a deadline, e.g. when there were technical difficulties during the exercise. In that case, it might have been possible that some students already started the practice mode. This should not be displayed after the deadline was postponed into the future again.

Additionally this PR fixes a bug where students started a programming exercise, but the repo was not set up fully, after which they could not interact with the exercise at all.

### Description

This PR hides any related things to the practice mode when it was started and the due date was postponed afterwards. In particular the graded repository is selected on default in the open repo and clone repo button before the new due date and resuming the practice participation is not possible when the due date is in the future.

Additionally when the started participation is not correctly set up (i.e. its status is not initialized), the student gets an error message and can try to restart the exercise.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise ideally with an old practice participation

1. Log in to Artemis
2. Start a practice participation after the due date
3. Set the due date in the future
4. No longer see the option to resume the practice participation (if you have an old practice participation at hand for testing)
5. Have the graded participation chosen as default in the clone/open repo button
6. The changes for failed setups is not really testable...


### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| clone-repo-button.component.ts | 80% | 90% |
| start-practice-mode-button.component.ts | 80% | 92% |
| exercise.utils.ts | 89% | 88% |
